spark/src/test/java/org/apache/iceberg/actions/TestExpireSnapshotsAction.java

```diff
@@ -0,0 +1,780 @@
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=2") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+  static final DataFile FILE_D = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-d.parquet")
+      .withFileSizeInBytes(10)
+      .withPartitionPath("c1=3") // easy way to set partition data for now
+      .withRecordCount(1)
+      .build();
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private File tableDir;
+  private String tableLocation;
+  private Table table;
+
+  @Before
+  public void setupTableLocation() throws Exception {
+    this.tableDir = temp.newFolder();
+    this.tableLocation = tableDir.toURI().toString();
+    this.table = TABLES.create(SCHEMA, SPEC, Maps.newHashMap(), tableLocation);
+  }
+
+  private Dataset<Row> buildDF(List<ThreeColumnRecord> records) {
+    return spark.createDataFrame(records, ThreeColumnRecord.class).coalesce(1);
+  }
+
+  private void writeDF(Dataset<Row> df, String mode) {
+    df.select("c1", "c2", "c3")
+        .write()
+        .format("iceberg")
+        .mode(mode)
+        .save(tableLocation);
+  }
+
+  private void checkExpirationResults(Long expectedDatafiles, Long expectedManifestsDeleted,
+      Long expectedManifestListsDeleted, ExpireSnapshotsActionResult results) {
+
+    Assert.assertEquals("Incorrect number of manifest files deleted",
+        expectedManifestsDeleted, results.getManifestFilesDeleted());
+    Assert.assertEquals("Incorrect number of datafiles deleted",
+        expectedDatafiles, results.getDataFilesDeleted());
+    Assert.assertEquals("Incorrect number of manifest lists deleted",
+        expectedManifestListsDeleted, results.getManifestListsDeleted());
+  }
+
+  @Test
+  public void testFilesCleaned() throws Exception {
+    Dataset<Row> df = buildDF(RECORDS);
+
+    writeDF(df, "append");
+
+    List<Path> expiredDataFiles = Files
+        .list(tableDir.toPath().resolve("data").resolve("c1=1"))
+        .collect(Collectors.toList());
+
+    Assert.assertEquals("There should be a data file to delete but there was none.",
+        2, expiredDataFiles.size());
+
+    writeDF(df, "overwrite");
+    writeDF(df, "append");
+
+    long end = System.currentTimeMillis();
+    while (end <= table.currentSnapshot().timestampMillis()) {
+      end = System.currentTimeMillis();
+    }
+
+    ExpireSnapshotsActionResult results =
+        Actions.forTable(table).expireSnapshots().expireOlderThan(end).execute();
+
+    table.refresh();
+
+    Assert.assertEquals("Table does not have 1 snapshot after expiration", 1, Iterables.size(table.snapshots()));
+
+    for (Path p : expiredDataFiles) {
+      Assert.assertFalse(String.format("File %s still exists but should have been deleted", p),
+          Files.exists(p));
+    }
+
+    checkExpirationResults(1L, 2L, 2L, results);
+  }
+
+  @Test
+  public void dataFilesCleanupWithParallelTasks() throws IOException {
```

Comment:
Ported test - everything the same except for the manifest manipulation see comment below